src/vm/diagnosticsprotocol.h

```diff
@@ -49,6 +55,486 @@ bool TryParseString(uint8_t *&bufferCursor, uint32_t &bufferLen, const T *&resul
+
+            return m_pData != NULL;
+        };
+
+        // Attempt to populate header and payload from a buffer.
+        // Payload is left opaque as a flattened buffer in m_pData
+        bool TryParse(::IpcStream* pStream)
+        {
+            CONTRACTL
+            {
+                NOTHROW;
+                GC_TRIGGERS;
+                MODE_PREEMPTIVE;
+                PRECONDITION(pStream != nullptr);
+            }
+            CONTRACTL_END;
+
+            // Read out header first
+            uint32_t nBytesRead;
+            bool success = pStream->Read(&m_Header, sizeof(IpcHeader), nBytesRead);
+            if (!success || nBytesRead < sizeof(IpcHeader))
+            {
+                return false;
+            }
+
+            m_Size = m_Header.Size;
+
+            // Then read out payload to buffer
+            uint16_t payloadSize = m_Header.Size - sizeof(IpcHeader);
+            if (payloadSize != 0)
+            {
+                BYTE* temp_buffer = new (nothrow) BYTE[payloadSize];
+                if (temp_buffer == nullptr)
+                {
+                    // OOM
+                    return false;
+                }
+
+                success = pStream->Read(temp_buffer, payloadSize, nBytesRead);
+                if (!success || nBytesRead < payloadSize)
+                {
+                    delete[] temp_buffer;
+                    return false;
+                }
+                m_pData = temp_buffer;
+            }
+
+            return true;
+        };
+
+        // Create a buffer of the correct size filled with
+        // header + payload. Correctly handles flattening of
+        // trivial structures, but uses a bool(Flatten)(void*)
+        // and uint16_t(GetSize)() when available.
+
+        // Handles the case where the payload structure exposes Flatten
+        // and GetSize methods
+        template <typename U,
+                  typename std::enable_if<HasFlatten<U>::value&& HasGetSize<U>::value, int>::type = 0>
+        bool FlattenImpl(U& payload)
+        {
+            CONTRACTL
+            {
+                NOTHROW;
+                GC_TRIGGERS;
+                MODE_PREEMPTIVE;
+            }
+            CONTRACTL_END;
+
+            if (IsFlattened())
+                return true;
+
+            S_UINT16 temp_size = S_UINT16(0);
+            temp_size += sizeof(struct IpcHeader) + payload.GetSize();
+            ASSERT(!temp_size.IsOverflow());
+
+            m_Size = temp_size.Value();
+
+            BYTE* temp_buffer = new (nothrow) BYTE[m_Size];
+            if (temp_buffer == nullptr)
+            {
+                // OOM
+                return false;
+            }
+
+            BYTE* temp_buffer_cursor = temp_buffer;
+
+            m_Header.Size = m_Size;
+
+            memcpy(temp_buffer_cursor, &m_Header, sizeof(struct IpcHeader));
+            temp_buffer_cursor += sizeof(struct IpcHeader);
+
+            payload.Flatten(temp_buffer_cursor);
+
```

Comment:
Consider adding an `_ASSERTE(m_pData == nullptr)` right before the assignment, that makes the code easier to understand that this is not going to cause a memory leak.